osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecurityKeyService.java

```diff
@@ -1,30 +1,96 @@
 import org.opensmartgridplatform.shared.exceptionhandling.FunctionalException;
 
 public interface SecurityKeyService {
-    byte[] reEncryptKey(final byte[] externallyEncryptedKey, final SecurityKeyType keyType) throws FunctionalException;
-    byte[] decryptKey(final byte[] encryptedKey, final SecurityKeyType keyType) throws ProtocolAdapterException;
-    byte[] encryptKey(final byte[] plainKey, final SecurityKeyType keyType) throws ProtocolAdapterException;
-    byte[] getDlmsMasterKey(final String deviceIdentification);
-    byte[] getDlmsAuthenticationKey(final String deviceIdentification);
-    byte[] getDlmsGlobalUnicastEncryptionKey(final String deviceIdentification);
-    byte[] getMbusDefaultKey(final String mbusDeviceIdentification);
-    byte[] getMbusUserKey(final String mbusDeviceIdentification);
-    byte[] getDlmsPassword(final String deviceIdentification);
-    DlmsDevice storeNewKey(final DlmsDevice device, final byte[] encryptedKey, final SecurityKeyType keyType);
-    DlmsDevice validateNewKey(final DlmsDevice device, final SecurityKeyType keyType) throws ProtocolAdapterException;
-    byte[] generateKey();
-    byte[] generateAndEncryptKey();
-    byte[] encryptMbusUserKey(final byte[] mbusDefaultKey, final byte[] mbusUserKey) throws ProtocolAdapterException;
+
+    byte[] getKey(String deviceIdentification, SecurityKeyType type);
+    Map<SecurityKeyType, byte[]> getKeys(String deviceIdentification, List<SecurityKeyType> keyTypes);
+
+    /**
+     * Store new key
+     * <p>
+     * A new key is a security key with a device which status NEW.
+     * This status is used when the new key is known, but not yet set on the device.
+     * <p>
+     * <strong>CAUTION:</strong> Only call this method when a successful
+     * connection with the device has been set up (that is: a valid
+     * communication key that works is known), and you are sure any existing new
+     * key data that is not activated yet (for instance a new key stored earlier in an
+     * attempt to replace the communication key that got aborted).<br>
+     * <p>
+     * The moment the new key is known to be transferred to the device, make
+     * sure to activate it by calling
+     * {@link #activateNewKey(String, SecurityKeyType)}.
+     *
+     * @param deviceIdentification
+     *         DLMS device id
+     * @param key
+     *        key to store, unencrypted
+     * @param keyType
+     *         type of key
+     *
+     *
+     * @see #activateNewKey(String, SecurityKeyType)
+     */
+    void storeNewKey(final String deviceIdentification, final SecurityKeyType keyType, final byte[] key);
+    void storeNewKeys(final String deviceIdentification, final Map<SecurityKeyType, byte[]> keysByType);
+
+    /**
+     * @see #storeNewKey(String, SecurityKeyType, byte[])
+     */
+    //void aesDecryptAndStoreNewKey(final String deviceIdentification, final SecurityKeyType keyTypes,
+    //        final byte[] aesEncryptedKeys) throws FunctionalException;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)